### PR TITLE
allow multiple args to `dbg!(..)`

### LINKED ISSUE #59763: Suggest parenthesising multiple arguments to dbg!()
```rust
fn main() {
    dbg!(5, 10); // error: no rules expected the token `,`
}
```
It would be nice if the compiler suggested the intended form: `dbg!((5, 10))` as this seems like a reasonable mistake. This is specific to a particular macro, though, so I'm not sure how flexible we can be with diagnostics here.

### PR DESCRIPTION
**Pull request by llogiq**
This closes #59763

### CODE REVIEW

File: src/libstd/macros.rs [hunk 1]
```diff
@@ -333,6 +340,9 @@ macro_rules! dbg {
                 tmp
             }
         }
+    };
+    ($val:expr, $($more:expr),+) => {
```
Comment (Centril) [inline src/libstd/macros.rs]:
```suggestion
    ($($val:expr),+ $(,)?) => {
```

File: src/libstd/macros.rs [hunk 2]
```diff
@@ -333,6 +340,9 @@ macro_rules! dbg {
...
+        dbg!(($val, $($more),*))
```
Comment (Centril) [inline src/libstd/macros.rs]:
```suggestion
        ($(dbg!($val)),+,)
```

**Review by Centril — COMMENTED** [review]
I think a better implementation of this, as I originally envisaged, was to have each part of the tuple on a separate line for better readability. Moreover, `dbg!(1,)` and `dbg!(2, 3,)` is supported by this change whereas it is not by your implementation.

**Comment by Centril** [comment]
Moreover, the UI tests for `dbg!` should be extended to cover the new behavior once we settle on a behavior, assuming we want to make this change.

File: src/test/ui/rfc-2361-dbg-macro/dbg-macro-expected-behavior.rs [hunk 3]
```diff
@@ -54,6 +54,17 @@ fn test() {
         7331
     }));
     assert_eq!(foo, 42);
+
+    // Test trailing comma:
+    assert_eq!("Yeah", dbg!("Yeah", ));
```
Comment (Centril) [inline src/test/ui/rfc-2361-dbg-macro/dbg-macro-expected-behavior.rs]: The result of `dbg!(x,)` is of type `(typeof(x),)`. IOW, a 1-element tuple is produced.

Comment (llogiq) [reply src/test/ui/rfc-2361-dbg-macro/dbg-macro-expected-behavior.rs]: Yep, I just changed the test to that effect.

**Comment by alexcrichton** [comment]
r? @SimonSapin 

I think you authored the original RFC, so you're probably best to look at this

**Comment by SimonSapin** [comment]
In the RFC, I specifically proposed *not* doing this:

https://github.com/rust-lang/rfcs/blob/master/text/2361-dbg-macro.md#accepting-a-single-expression-instead-of-many

I think it creates an undesirable “discontinuity”: if `dbg!(x, y, z)` returns the 3-tuple `(x, y, y)` and `dbg!(x, y)` returns the 2-tuple `(x, y)` (and `dbg!()` returns the zero-tuple `()`) then the consistent continuation is that `dbg!(x)` returns the 1-tuple `(x,)`. But it has been stabilized as returning `x`, and a 1-tuple is probably less convenient anyway.

I can’t find it again, but I vaguely remember an argument that you could look at `dbg!(x)` as `dbg!  (x)`, and `(x)` by itself is `x` with unnecessary parens, not a tuple. But I disagree with this framing: the `foo!(…)` syntax of expression-position macro invocation is much more similar to function call syntax than to tuple syntax. (Using a 1-tuple to represent the arguments of a single-argument function appears again in the desugared notation for the `Fn(T)` trait, which is `Fn<(T,)>`.) 

This PR seems to be going for emulating tuple syntax: it makes `dbg!(x,)` behave differently than `dbg!(x)` and return `(x,)` instead of `x`. But in a function call, a trailing comma after arguments is not significant. I think we should not make it significant in a function-like macro that takes comma-separated expressions as arguments.

----

**TL;DR:** I feel I would still prefer that `dbg!(…)` only accepts one (or zero) expression, and that making it return a tuple involves passing it an actual tuple value.

That said, is there consensus except for me that the convenience of debugging multiple values with fewer sigils is worth accepting this weird inconsistency?

If so, I’d prefer we embrace the inconsistency and make `dbg!(x,)` the same as `dbg!(x)`. And maybe the filename and line number should not be repeated for every element of the tuple?

CC @rust-lang/libs

**Comment by Centril** [comment]
> I can’t find it again, but I vaguely remember an argument that you could look at `dbg!(x)` as `dbg! (x)`, and `(x)` by itself is `x` with unnecessary parens, not a tuple.

Yes, that is the framing I was going for. It provides a good internal consistency. Perhaps not the one you want, but I think it works well in practice to do what the users wants 99% of the time.

> This PR seems to be going for emulating tuple syntax: it makes `dbg!(x,)` behave differently than `dbg!(x)` and return `(x,)` instead of `x`. But in a function call, a trailing comma after arguments is not significant. I think we should not make it significant in a function-like macro that takes comma-separated expressions as arguments.

What is the failure mode here? i.e. in what way do you think this will trip users up?

Note that if you write `dbg!(x);` or `dbg!(x,);` it will not matter that you actually received back a tuple as the semicolon turns the type into `()` in any case.

To actually have a problem, you need to actually use the result:

```rust
let x = dbg!(x,);
```
(`x` must still be passed along to have a problem)

I think this is unlikely and is only probable as the result of a typo. When it happens the compiler already gives a good error message:

```rust
fn main() {
    let _: u8 = dbg!(1,);
}
```
results in:
```rust
error[E0308]: mismatched types
  --> src/main.rs:18:9
   |
18 |         ($(dbg!($val)),+,)
   |         ^^^^^^^^^^^^^^^^^^ expected u8, found tuple
...
23 |     let _: u8 = dbg!(1,);
   |                 -------- in this macro invocation
   |
   = note: expected type `u8`
              found type `({integer},)`
```

> If so, I’d prefer we embrace the inconsistency and make `dbg!(x,)` the same as `dbg!(x)`. And maybe the filename and line number should not be repeated for every element of the tuple?

Note that this only makes the macro more complex, not less. One good advantage of this PR is that it provides a simple implementation that should also be simple to explain: simply remove `dbg!` and the thing you have without is the value you get back out and of that type.

**Comment by BurntSushi** [comment]
If this PR is accepted, is the type of `dbg!(a, b)` and `dbg!((a, b))` the same?

I do agree with @SimonSapin that if we're going to allow multiple arguments, then the trailing comma in the case of the first argument should behave like a function call. While `let x = dbg!(y,);` is unlikely to happen in practice, something like this is much more feasible:

```rust
let x = dbg!(
    some_long_single_argument_expression,
);
```

And if that doesn't work---regardless of whether we permit multiple arguments or not---I would honestly be pretty surprised. I'd expect things like this to work as expected as well:

```rust
let x = dbg!(
    some_long_single_argument_expression,
    another_long_single_argument_expression,
);
```

I don't have too many strong opinions on whether we permit multiple arguments. It does honestly seem nice, and the case of _wanting_ a unit tuple is a bit weird/rare anyway, so if you did need that, you'd just have to get used to writing `dbg!((x,))` I guess.

**Comment by Centril** [comment]
> If this PR is accepted, is the type of `dbg!(a, b)` and `dbg!((a, b))` the same?

Yes. (And `(a, b)` and `((a, b))` are of the same type.)

**Comment by SimonSapin** [comment]
>> If so, I’d prefer we embrace the inconsistency and make dbg!(x,) the same as dbg!(x).
>
> Note that this only makes the macro more complex, not less.

IMO the complexity is of the implementation is much much less important than the complexity of the documentation that accurately explains the behavior. (aka “how we teach this”)

> One good advantage of this PR is that it provides a simple implementation that should also be simple to explain: simply remove dbg! and the thing you have without is the value you get back out and of that type.

I feel strongly that we should *not* document it this way. Macros that take comma-separated expressions look like function calls (even though they can do things functions can’t). I think going against that flow would be harmful for teachability.

**Comment by Centril** [comment]
> IMO the complexity is of the implementation is much much less important than the complexity of the documentation that accurately explains the behavior. (aka “how we teach this”)

As the implementation gives rise to a simple rule, I think the complexity of the documentation will also be simple.

**Comment by scottmcm** [comment]
While I think I originally liked the multi-argument version, I'm perfectly fine with it always taking one argument.  It's not materially harder to type `dbg!((a, b))` than `dbg!(a, b)`, and there are plenty of other places where I end up doubling parens like that -- `heap.push((a, b))`, for example.

**Comment by Centril** [comment]
@scottmcm It's not that I find that `dbg!((a, b))` is harder to type, I don't. I implemented `dbg!` yet I still mistakenly try to type `dbg!(a, b, ...)` every once in a while when opening the playground. This results in an error and I can easily fix it. However, this causes friction and a slowdown because I have to recompile. If it causes problems for the person who implemented `dbg!` then it stands to reason that beginners and novices will have a harder time.

**Comment by llogiq** [comment]
I'm thinking about changing the definition to a proc macro that would allow us to use the line and column numbers of the debugged expressions instead of the macro call. This might make it easier to read the output.

**Comment by SimonSapin** [comment]
The libs team discussed this in today’s triage meeting. Consensus was to:

* Make the macro accept multiple arguments, returning a tuple, despite the discontinuity with `dbg!(x)` not returning a 1-tuple. (Not that a 1-tuple would be desirable.)
* Document separately the three cases (zero argument, one argument, and 2+ arguments) and respective return types.
* Make the macro accept a trailing comma (except for the zero-argument case) which is never significant, like in function call syntax. In particular, `dbg!(x)` returns `x`, not `(x,)`.
* Document the macro’s syntax as arguments (which evokes a function call), not in terms of tuple syntax with a `dbg!` prefix.

@llogiq, could you update the PR accordingly?

**Comment by Centril** [comment]
> * In particular, `dbg!(x)` returns `x`, not `(x,)`.

@SimonSapin You mean `dbg!(x,)` here, right?

**Comment by llogiq** [comment]
I read this as `dbg!(x,) == dbg!(x)` (yes, that means for 1-tuples we need double braces).

**Comment by SimonSapin** [comment]
Ah yes, that was a typo. Fixed.

**Comment by SimonSapin** [comment]
I’m personally still undecided (and forgot to bring it up in the meeting) whether it’s better if the file name and line number are printed once per macro call or once per argument.

**Comment by Centril** [comment]
@SimonSapin Since that's not part of the stable guarantees it seems to me that we can do the simple thing now and iterate / "feel" our way through.

**Comment by SimonSapin** [comment]
Sure, it’s not set in stone and we can change our minds later. I think it’s still worth discussing now, though! Do you have thoughts either way?

**Comment by Centril** [comment]
> Do you have thoughts either way?

It's tricky... on the one hand, having output like:

```
[src/main.rs:2] a = 1
[src/main.rs:2] b = 2
[src/main.rs:2] c = 3
```

feels nicely aligned... on the other it feels repetitive.

A sketch for an alternative for multi-argument output:

```
[src/main.rs:2]
[ a = 1
[ b = 2
[ c = 3
```

**Comment by llogiq** [comment]
I've implemented the libs team's suggestions as @SimonSapin wrote them up. I did not yet change the output of the macro, we can iterate on that later.

**Comment by llogiq** [comment]
@SimonSapin if you are OK with this I'd like to land this as is now and iterate on the output in a follow-up PR.

**Comment by SimonSapin** [comment]
Looks good, thanks!

@bors r+
